extensions/windows-event-log/wel/JSONUtils.cpp

```diff
@@ -0,0 +1,170 @@
+    system.AddMember("Task", rapidjson::StringRef(system_xml.child("Task").text().get()), doc.GetAllocator());
+    system.AddMember("Opcode", rapidjson::StringRef(system_xml.child("Opcode").text().get()), doc.GetAllocator());
+    system.AddMember("Keywords", rapidjson::StringRef(system_xml.child("Keywords").text().get()), doc.GetAllocator());
+
+    {
+      auto timeCreated_xml = system_xml.child("TimeCreated");
+      auto& timeCreated = flatten ? doc : system.AddMember("TimeCreated", rapidjson::kObjectType, doc.GetAllocator())["TimeCreated"];
+      timeCreated.AddMember("SystemTime", rapidjson::StringRef(timeCreated_xml.attribute("SystemTime").value()), doc.GetAllocator());
+    }
+
+    system.AddMember("EventRecordID", rapidjson::StringRef(system_xml.child("EventRecordID").text().get()), doc.GetAllocator());
+
+    {
+      auto correlation_xml = system_xml.child("Correlation");
+      auto& correlation = flatten ? doc : system.AddMember("Correlation", rapidjson::kObjectType, doc.GetAllocator())["Correlation"];
+      correlation.AddMember("ActivityID", rapidjson::StringRef(correlation_xml.attribute("ActivityID").value()), doc.GetAllocator());
+    }
+
+    {
+      auto execution_xml = system_xml.child("Execution");
+      auto& execution = flatten ? doc : system.AddMember("Execution", rapidjson::kObjectType, doc.GetAllocator())["Execution"];
+      execution.AddMember("ProcessID", rapidjson::StringRef(execution_xml.attribute("ProcessID").value()), doc.GetAllocator());
+      execution.AddMember("ThreadID", rapidjson::StringRef(execution_xml.attribute("ThreadID").value()), doc.GetAllocator());
+    }
+
+    system.AddMember("Channel", rapidjson::StringRef(system_xml.child("Channel").text().get()), doc.GetAllocator());
+    system.AddMember("Computer", rapidjson::StringRef(system_xml.child("Computer").text().get()), doc.GetAllocator());
+  }
+
+  {
+    auto eventData_xml = event_xml.child("EventData");
+    // create EventData subarray even if flatten requested
+    doc.AddMember("EventData", rapidjson::kArrayType, doc.GetAllocator());
+    for (const auto& data : eventData_xml.children()) {
+      auto name_attr = data.attribute("Name");
+      rapidjson::Value item(rapidjson::kObjectType);
+      item.AddMember("Name", rapidjson::StringRef(name_attr.value()), doc.GetAllocator());
+      item.AddMember("Content", rapidjson::StringRef(data.text().get()), doc.GetAllocator());
+      item.AddMember("Type", rapidjson::StringRef(data.name()), doc.GetAllocator());
+      // we need to query EventData because a reference to it wouldn't be stable, as we
+      // possibly add members to its parent which could result in reallocation
+      doc["EventData"].PushBack(item, doc.GetAllocator());
+      // check collision
+      if (flatten && !name_attr.empty() && !doc.HasMember(name_attr.value())) {
+        doc.AddMember(rapidjson::StringRef(name_attr.value()), rapidjson::StringRef(data.text().get()), doc.GetAllocator());
```

Comment:
If I understand correctly, using `rapidjson::StringRef` all over the place means that the returned json object is valid only as long as the original pugi xml object is valid and unchanged. If so, there should be a big loud warning on the function to warn the next caller about this.

Reply:
added big loud warning 🙂